src/org/jetbrains/java/decompiler/modules/decompiler/TryWithResourcesProcessor.java

```diff
@@ -348,14 +351,21 @@ public static void findEdgesLeaving(Statement curr, Statement check, Set<StatEdg
   }
 
   public static void findEdgesLeaving(Statement curr, Statement check, Set<StatEdge> edges, boolean allowExit) {
+    findEdgesLeaving(curr, check, edges, allowExit, new HashMap<>());
+  }
+
+  private static void findEdgesLeaving(Statement curr, Statement check, Set<StatEdge> edges, boolean allowExit, Map<Statement, Set<Statement>> found) {
     for (StatEdge edge : curr.getAllSuccessorEdges()) {
       if (!check.containsStatement(edge.getDestination()) && (allowExit || !(edge.getDestination() instanceof DummyExitStatement))) {
-        edges.add(edge);
+        if (edge.explicit || found.entrySet().stream().allMatch(e -> !e.getKey().containsStatement(curr) || !e.getValue().contains(edge.getDestination()))) {
```

Comment:
If you could add a comment here as to what the check is doing, that'd be great :)